index.rb

```diff
@@ -1,160 +1,235 @@
-      end
+    each do |i|
+      my_count_counter += 1 if i == my_parameter
     end
-    puts my_count_counter
+    my_count_counter
   end
 
-  def my_map
-
+  def my_map_prock(block)
     my_array = []
-    
-    #Replace the .each with my_each
 
-    for i in self
-      my_array.push(yield(i))
+    each do |i|
+      my_array.push(block.call(i))
     end
     my_array
-
   end
 
+  def my_map(block = nil)
+    return to_enum(:my_map) unless block_given?
+    return my_map_prock(block) unless block.nil?
 
-  def my_inject
+    my_array = []
 
-    my_aggregator = self.first
-    
-    for i in self do
-      if i == self.first then
-        next
-      end
-      my_aggregator = yield(my_aggregator, i)
+    each do |i|
+      my_array.push(yield(i))
     end
-
-    puts my_aggregator
-
+    my_array
   end
 
-  def self.multiply_els(my_array)
+  def my_inject(*args)
+    return my_inject { |n, i| n.send(args.first, i) } if args.length == 1 && !block_given?
 
-    my_array.my_inject {|a,b| a * b}
-  end
+    if args.length == 2 && !block_given?
+      my_aggregator = args.first
 
+      each do |i|
+        my_aggregator = my_aggregator.send(args[1], i)
+      end
 
-end
+      return my_aggregator
+    end
+
+    if args.length == 1
+      my_aggregator = args.first
 
+      each do |i|
+        my_aggregator = yield(my_aggregator, i)
+      end
 
-#Testing all the codes,  ADD YOUR TESTS HERE
-[1,2,3].my_each { |n| puts n * 3}
-[1,2,3].my_each_with_index { |number, index| puts "the number: #{number}, has index: #{index}"}
-[1,2,3,4].my_select { |n| n.even?}
-[0,2,4,8].my_all? {|n| n.even?}
-[1,3,5,7].my_any? {|n| n.even?}
-[1,3,5,7].my_none? {|n| n.even?}
-[1,3,5,7,1,nil,5,7,1,nil,5,7].my_count
-[1,3,5,7,1,nil,5,7,1,nil,5,7].my_count(1)
-some_new_array = [1,3,4,8].my_map { |n| n * 3}
-some_other_new_array = [1,3,4,8].my_each {|n| n  * 3}
-puts some_new_array
-puts some_other_new_array
-[1,3,5,7].my_inject { |a,b| a + b}
+      return my_aggregator
+    end
 
+    my_aggregator = first
 
-movie = Proc.new {return "something"}
-puts movie.call
+    each do |i|
+      next if i == first
 
-Enumerable.multiply_els([1,3,4,7,8,12])
+      my_aggregator = yield(my_aggregator, i)
+    end
 
+    my_aggregator
+  end
 
-  # https://medium.com/rubycademy/the-yield-keyword-603a850b8921
+  def multiply_els()
+    my_inject { |a, b| a * b }
+  end
+end
 
-  # https://jsonmattingly.medium.com/writing-your-own-enumerables-in-ruby-18db4fa0e5fe
\ No newline at end of file
+# array = [1,2,3,4,5]
+# numbers = array
+# operation = Proc.new {|n| n + 1}
+
+# puts array.my_all?(Integer) ==  array.all?(Integer)
+# puts numbers.my_all?(Numeric) == numbers.all?(Numeric)
+# puts array.my_all?(3) == array.all?(3)
+# puts array.my_any?(Numeric) == array.any?(Numeric)
+# puts array.my_inject(&operation) == array.inject(&operation)
+# puts array.my_inject(:+) == array.inject(:+)
+# puts array.multiply_els()
+# puts "fix multiply_els to be just a method"
```

Comment:
From line 224 to line 235 are debugging code. kindly delete or keep for yourself on another branch. :hammer_and_wrench: